[inductor] Add shape checks to ExpandView

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113839

Currently `ExpandView` doesn't check that the expanded shape is valid which may
allow bugs to slip through which cause silent correctness issues.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

peterbell10 committed:
[inductor] Add shape checks to ExpandView

Currently `ExpandView` doesn't check that the expanded shape is valid which may
allow bugs to slip through which cause silent correctness issues.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Add shape checks to ExpandView"


Currently `ExpandView` doesn't check that the expanded shape is valid which may
allow bugs to slip through which cause silent correctness issues.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 committed:
Lint on "[inductor] Add shape checks to ExpandView"


Currently `ExpandView` doesn't check that the expanded shape is valid which may
allow bugs to slip through which cause silent correctness issues.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ezyang commented on torch/_inductor/ir.py:
```diff
@@ -1680,10 +1680,20 @@ def _normalize_size(x, new_size):
+            elif is_unbacked(new_size[i]):
+                # Assume equal, take backed symint
+                # TODO: This may be unsound
+                new_size[i] = old_size[i]
```
You want to make a deferred runtime assert. You can trigger that with expect_true(new_size[i] == old_size[i]). Actually it's a bit strange `guard_equals` doesn't do this automatically

ezyang commented:
Do you think you can easily construct a test?

peterbell10 committed:
Update on "[inductor] Add shape checks to ExpandView"


Currently `ExpandView` doesn't check that the expanded shape is valid which may
allow bugs to slip through which cause silent correctness issues.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Add shape checks to ExpandView"


Currently `ExpandView` doesn't check that the expanded shape is valid which may
allow bugs to slip through which cause silent correctness issues.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Add shape checks to ExpandView"


Currently `ExpandView` doesn't check that the expanded shape is valid which may
allow bugs to slip through which cause silent correctness issues.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Add shape checks to ExpandView"


Currently `ExpandView` doesn't check that the expanded shape is valid which may
allow bugs to slip through which cause silent correctness issues.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Add shape checks to ExpandView"


Currently `ExpandView` doesn't check that the expanded shape is valid which may
allow bugs to slip through which cause silent correctness issues.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ezyang reviewed (approved):
ok. I'm not entirely sure how we will keep using these new sizevars functions but we can figure it out as we go

peterbell10 committed:
Update on "[inductor] Add shape checks to ExpandView"


Currently `ExpandView` doesn't check that the expanded shape is valid which may
allow bugs to slip through which cause silent correctness issues.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge

peterbell10 commented:
@pytorchbot merge

ezyang commented on test/inductor/test_unbacked_symints.py:
```diff
@@ -30,6 +31,16 @@ def fn(x, y):
+    def test_expand_mismatch(self):
+        def fn(x):
+            nz = x.nonzero()
+            return nz.expand([-1, 128])
```
I was looking over this test, and after some of my recent changes, it doesn't actually exercise the relevant Inductor codepath. Specifically, it instead fails with:

```
Failed running call_method expand(*(FakeTensor(..., device='cuda:0', size=(u0, 2), dtype=torch.int64), [-1, 128]), **{}): expand: attempting to expand a dimension of length 2!
```

ezyang replied:
I can try to modify it as 

```
diff --git a/test/inductor/test_unbacked_symints.py b/test/inductor/test_unbacked_symints.py
index 8d7228e0d4c..1bdc95b7c0d 100644
--- a/test/inductor/test_unbacked_symints.py
+++ b/test/inductor/test_unbacked_symints.py
@@ -41,7 +41,7 @@ class TestUnbackedSymints(InductorTestCase):
     def test_expand_mismatch(self, device):
         def fn(x):
             nz = x.nonzero()
-            return nz.expand([-1, 128])
+            return nz.expand([-1, 128, 2])
 
         x = make_tensor(32, 4, device=device, dtype=torch.float32, exclude_zero=True)
         with self.assertRaises(torch._dynamo.exc.TorchRuntimeError):
```

but this also fails early with

```
torch._dynamo.exc.UserError: Tried to use data-dependent value in the subsequent computation. This can happen when we encounter unbounded dynamic value that
 is unknown during tracing time.  You will need to explicitly give hint to the compiler. Please take a look at constrain_as_value OR constrain_as_size APIs.
  Could not guard on data-dependent expression Eq(u0, 128) (unhinted: Eq(u0, 128)).  (Size-like symbols: u0) 
```

So I'm not sure how we can get to the Inductor code without having already tested the necessary conditions.
